plugins/ldap/server/__init__.py

```diff
@@ -0,0 +1,204 @@
+
+
+def _registerLdapUser(attrs, email, server):
+    first, last = None, None
+    if attrs.get('givenName'):
+        first = attrs['givenName'][0]
+    elif attrs.get('cn'):
+        first = attrs['cn'][0].split()[0]
+
+    if attrs.get('sn'):
+        last = attrs['sn'][0]
+    elif attrs.get('cn'):
+        last = attrs['cn'][0].split()[-1]
+
+    if not first or not last:
+        raise Exception('No LDAP name entry found for %s.' % email)
+
+    # Try using the search field value as the login. If it's an email address,
+    # use the part before the @.
+    try:
+        login = attrs[server['searchField']][0].split('@')[0]
+        return ModelImporter.model('user').createUser(
+            login, password=None, firstName=first, lastName=last, email=email)
+    except ValidationException as e:
+        if e.field != 'login':
+            raise
+
+    # Fall back to deriving login from user's name
+    for i in six.moves.range(_MAX_NAME_ATTEMPTS):
+        login = ''.join((first, last, str(i) if i else ''))
+        try:
+            return ModelImporter.model('user').createUser(
+                login, password=None, firstName=first, lastName=last, email=email)
+        except ValidationException as e:
+            if e.field != 'login':
+                raise
+
+    raise Exception('Failed to generate login name for LDAP user %s.' % email)
+
+
+def _getLdapUser(attrs, server):
+    emails = attrs.get('mail')
+    if not emails:
+        raise Exception('No email record present for the given LDAP user.')
+
+    if not isinstance(emails, (list, tuple)):
+        emails = (emails,)
+
+    userModel = ModelImporter.model('user')
+    for email in emails:
+        existing = userModel.findOne({
+            'email': {'$eq': email.lower()}
+        })
+        if existing:
+            return existing
+
+    return _registerLdapUser(attrs, email, server)
+
+
+def _ldapAuth(event):
+    login, password = event.info['login'], event.info['password']
+
+    # TODO cache LDAP servers list setting
+    servers = ModelImporter.model('setting').get(PluginSettings.LDAP_SERVERS)
+
+    for server in servers:
+        # ldap requires a uri complete with protocol.
+        # Append one if the user did not specify.
+        uri = server['uri']
+        if '://' not in uri:
+            uri = 'ldap://' + uri
+        conn = ldap.initialize(uri)
+        conn.set_option(ldap.OPT_TIMEOUT, _CONNECT_TIMEOUT)
+        conn.set_option(ldap.OPT_NETWORK_TIMEOUT, _CONNECT_TIMEOUT)
+
+        try:
+            conn.bind_s(server['bindName'], server['password'], ldap.AUTH_SIMPLE)
+        except ldap.LDAPError:
+            logger.exception('LDAP connection failed (%s).' % server['uri'])
+            continue
+
+        searchStr = '%s=%s' % (server['searchField'], login)
+        results = conn.search_s(server['baseDn'], ldap.SCOPE_ONELEVEL, searchStr, _LDAP_ATTRS)
+
+        if results:
+            entry, attrs = results[0]
```

Comment:
We are assuming that the searchField should identify at most one entry; if it doesn't, that is a problem that we should probably log.